### PR TITLE
refactor Comparables 

### PR DESCRIPTION
**Pull request by zhexuany**
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tikv-client-lib-java/131)

### CODE REVIEW

File: src/test/java/com/pingcap/tikv/predicates/PredicateUtilsTest.java [hunk 1]
```diff
@@ -29,19 +29,19 @@
   public void mergeCNFExpressions() throws Exception {
     List<TiExpr> exprs =
         ImmutableList.of(
-            TiConstant.create(1),
-            TiConstant.create(2),
-            TiConstant.create(3),
-            TiConstant.create(4),
-            TiConstant.create(5));
+            new TiConstant(1),
```
Comment (ilovesoup) [inline src/test/java/com/pingcap/tikv/predicates/PredicateUtilsTest.java:32]: what's the purpose to replace static method with new?

File: src/main/java/com/pingcap/tikv/expression/TiConstant.java [hunk 2]
```diff
@@ -56,49 +64,23 @@ public Object getValue() {
...
-    } else if (value instanceof String) {
-      return DataTypeFactory.of(TYPE_VARCHAR);
-    } else if (value instanceof Float) {
-      return DataTypeFactory.of(TYPE_FLOAT);
-    } else if (value instanceof Double) {
-      return DataTypeFactory.of(TYPE_NEW_DECIMAL);
-    } else {
-      throw new TiExpressionException("Constant type not supported.");
     }
+    return value.getDataType();
```
Comment (birdstorm) [inline src/main/java/com/pingcap/tikv/expression/TiConstant.java:83]: I wonder what happens if the value is of `BigDecimal` type, will this be a potential bug here?

File: src/main/java/com/pingcap/tikv/meta/TiSelectRequest.java [hunk 3]
```diff
@@ -26,7 +26,6 @@
 import com.pingcap.tikv.expression.TiByItem;
 import com.pingcap.tikv.expression.TiColumnRef;
 import com.pingcap.tikv.expression.TiExpr;
-import com.pingcap.tikv.kvproto.Coprocessor.KeyRange;
```
Comment (birdstorm) [inline src/main/java/com/pingcap/tikv/meta/TiSelectRequest.java]: why is this import deleted?

File: src/main/java/com/pingcap/tikv/expression/TiConstant.java (hunk 2 above)
Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/expression/TiConstant.java:83]: Yeah. This what I am trying to prevent.

Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/expression/TiConstant.java:83]: More stuffs come out tonight(CDT)

File: src/main/java/com/pingcap/tikv/util/BytesComparable.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  private BytesComparable(Object obj) {
+    if(obj instanceof BytesComparable) {
+      throw new IllegalStateException("This object is already comparable");
+    }
+    this.value = convertObjectToBytes(obj);
+    this.obj = obj;
+  }
+
+  private byte[] convertObjectToBytes(Object o) {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/util/BytesComparable.java:52]: This looks completely wrong by design.
A comparableBytes means it's comparable in the same way as the order defined in TiDB type system. Once you convert them like this, the original order lost. You might get 1.1 > 2.2 or something like this stuff. This looks not right if they are not bytestring or byte.

**Review by ilovesoup — CHANGES_REQUESTED** [review]

File: src/main/java/com/pingcap/tikv/util/BytesComparable.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      in = new ObjectInputStream(bis);
+      o = in.readObject();
+    } catch (IOException | ClassNotFoundException ignored) {
+    }
+    return o;
+  }
+
+  @Override
+  public int compareTo(BytesComparable o) {
+    if(getObjectValue() instanceof ByteString) {
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/util/BytesComparable.java:79]: Use FastByteComparisons borrowed from Cassandra/Guava. That one is a lot faster.

**Comment by zhexuany** [comment]
@ilovesoup is working on this. Closing.

**Pull request closed without merging**